### ISSUE TITLE
Batch collation is slow/hangs when done outside DataLoader

### ISSUE THREAD

**Issue by jopo666** [pytorch/pytorch#85631]
### Bug:

Batching samples is fast when done inside `DataLoader`, and slower/hangs when done outside `DataLoader`. I am able to reproduce the bug with the following collation functions:
- `torch.utils.data.default_collate`
- `torch.stack`
- `tensor = torch.zeros(...)` and `tensor.copy_(...)` combo

The bug does not disappear if the `batched(loader)` function from the code below is replaced with a second `DataLoader`. To simplify the code below, I use `torch.stack` and only a single `DataLoader`.

### Code to reproduce the bug:
```python
import time

import torch
from torch.utils.data import DataLoader, Dataset

BATCH_SIZE = 64
SHAPE = (3, 224, 224)


class DebugDataset(Dataset):
    def __len__(self):
        return 6400

    def __getitem__(self, index):
        return torch.zeros(SHAPE)


def verbose_stack(samples):
    tic = time.perf_counter()
    batch = torch.stack(samples)
    print("tensor stacked: {:.3f} ms".format((time.perf_counter() - tic) * 1000))
    return batch

def batched(loader):
    """Helper function to collate batches outside the dataloader."""
    batch = []
    for sample in loader:
        batch.append(sample)
        if len(batch) >= BATCH_SIZE:
            # BUG: collation slow after the first data loader.
            # - the bug persists even when this collation is wrapped with a second data loader.
            yield verbose_stack(batch)
            batch = []


def main():
    print(" Fast ".center(40, "="))
    for images in DataLoader(
        DebugDataset(), num_workers=8, batch_size=64, collate_fn=verbose_stack
    ):
        assert images.shape[0] == BATCH_SIZE

    print(" Slow ".center(40, "="))
    for __ in batched(DataLoader(DebugDataset(), num_workers=8, batch_size=None)):
        assert images.shape[0] == BATCH_SIZE

if __name__ == "__main__":
    main()
```

### Output (truncated):

Stacking samples take approx. `10ms` when done inside the `DataLoader`. When done outside the `DataLoader` collation is a bit slower most of the time (`12ms`) but sometimes takes up to `500ms`.

```bash
================= Fast =================
tensor stacked: 68.234 ms
tensor stacked: 8.847 ms
tensor stacked: 7.620 ms
tensor stacked: 11.278 ms
tensor stacked: 9.667 ms
tensor stacked: 10.566 ms
tensor stacked: 9.321 ms
tensor stacked: 8.852 ms
tensor stacked: 8.447 ms
tensor stacked: 9.884 ms
tensor stacked: 8.907 ms
tensor stacked: 9.566 ms
tensor stacked: 9.449 ms
tensor stacked: 9.904 ms
tensor stacked: 10.012 ms
tensor stacked: 10.961 ms
tensor stacked: 7.505 ms
tensor stacked: 9.160 ms
tensor stacked: 8.321 ms
tensor stacked: 8.337 ms
tensor stacked: 10.346 ms
# and so on...
================= Slow =================
tensor stacked: 851.102 ms
tensor stacked: 30.045 ms
tensor stacked: 121.725 ms
tensor stacked: 12.526 ms
tensor stacked: 12.506 ms
tensor stacked: 473.693 ms
tensor stacked: 12.480 ms
tensor stacked: 12.424 ms
tensor stacked: 18.115 ms
tensor stacked: 112.905 ms
tensor stacked: 17.900 ms
tensor stacked: 12.541 ms
tensor stacked: 113.073 ms
tensor stacked: 18.081 ms
tensor stacked: 12.612 ms
tensor stacked: 113.725 ms
tensor stacked: 9.676 ms
tensor stacked: 12.551 ms
tensor stacked: 462.092 ms
tensor stacked: 12.598 ms
tensor stacked: 12.585 ms
tensor stacked: 18.121 ms
tensor stacked: 18.318 ms
tensor stacked: 12.104 ms
tensor stacked: 505.062 ms
# and so on...
```

### Why would you need to do this?

 My training data is split into tar shards with each shard containing samples from one label. As each batch from `DataLoader` contains samples from only one worker, batches might contain samples from only one label. To fix this, I need to shuffle each worker's outputs after the data loader (ie. `unbatch -> shuffle -> batch`). This is routine in [`webdataset`](https://github.com/webdataset/webdataset), where the bug could be reproduced with:

```python
from webdataset import WebDataset, WebLoader

dataset = WebDataset(url).shuffle(1000).batched(128)
slow_loader = WebLoader(dataset).unbatched().shuffle(1000).batched(128)
#                                                             ^
#                                                             |
#                                                    this would be slow
```

### Versions

```
Collecting environment information...
PyTorch version: 1.12.0+cu116
Is debug build: False
CUDA used to build PyTorch: 11.6
ROCM used to build PyTorch: N/A

OS: Rocky Linux 8.6 (Green Obsidian) (x86_64)
GCC version: (GCC) 9.2.1 20191120 (Red Hat 9.2.1-2)
Clang version: Could not collect
CMake version: version 3.20.2
Libc version: glibc-2.28

Python version: 3.9.7 (default, May 10 2022, 23:45:56)  [GCC 8.5.0 20210514 (Red Hat 8.5.0-10)] (64-bit runtime)
Python platform: Linux-4.18.0-372.19.1.el8_6.x86_64-x86_64-with-glibc2.28
Is CUDA available: True
CUDA runtime version: 11.6.124
GPU models and configuration: GPU 0: NVIDIA A100-SXM4-40GB
Nvidia driver version: 510.85.02
cuDNN version: Probably one of the following:
/usr/lib64/libcudnn.so.8.4.1
/usr/lib64/libcudnn_adv_infer.so.8.4.1
/usr/lib64/libcudnn_adv_train.so.8.4.1
/usr/lib64/libcudnn_cnn_infer.so.8.4.1
/usr/lib64/libcudnn_cnn_train.so.8.4.1
/usr/lib64/libcudnn_ops_infer.so.8.4.1
/usr/lib64/libcudnn_ops_train.so.8.4.1
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

Versions of relevant libraries:
[pip3] numpy==1.22.4
[pip3] pytorch-lightning==1.6.4
[pip3] pytorch-pfn-extras==0.5.8
[pip3] torch==1.12.0+cu116
[pip3] torch-tb-profiler==0.4.0
[pip3] torchaudio==0.12.0+cu116
[pip3] torchdata==0.4.0
[pip3] torchfile==0.1.0
[pip3] torchmetrics==0.9.2
[pip3] torchtext==0.13.0
[pip3] torchvision==0.13.0+cu116
[conda] Could not collect
```

**Comment by jopo666** [comment]
The slowdown was due to `sbatch` script settings, where `export OMP_NUM_THREADS=1` was not defined. Nothing to do with PyTorch.
